js_utils/text_validation.js

```diff
@@ -186,6 +186,65 @@ export class TextValidation {
     if (barcode_header !== "ML" && barcode_header !== "MS") {
       return " ";
     }
+    let barcode_err = "";
+    if (barcode.includes("-")) {
+      barcode_err = this._check_new_barcode(barcode, beta_2_mode);
+    } else {
+      barcode_err = this._check_old_barcode(barcode, beta_2_mode);
+    }
+    return barcode_err;
+  }
+
+  /**
+   * Returns the feedback text for the new plate barcode validation
+   *
+   * @param  {string}  barcode The barcode string to validate
+   * @param {bool} beta_2_mode True if in bet 2 mode false if in beta 1 mode
+   * @return {string} "" if barcode is valid, " " otherwise
+   *
+   */
+  _check_new_barcode(barcode, beta_2_mode) {
+    // check that charecters are alpha numeric exept the dash
+    const barcode_without_dash = barcode.slice(0, 10) + barcode[11];
+    const regEx = /^[0-9a-zA-Z]+$/;
+    if (!barcode_without_dash.match(regEx)) {
+      return " ";
+    }
+    // check if dash is in correct location
+    if (barcode[10] !== "-") {
+      return " ";
+    }
+    // check if the year is 2022 or later
+    if (parseInt(barcode.slice(2, 4)) < 22) {
+      return " ";
+    }
+    // check if the day is between 1 and 365 inclusive
+    if (parseInt(barcode.slice(4, 7)) < 1 || parseInt(barcode.slice(4, 7)) > 365) {
+      return " ";
+    }
+    // check that ### is between 0 and 299 inclusive
+    if (parseInt(barcode.slice(7, 10)) < 0 || parseInt(barcode.slice(7, 10)) > 299) {
+      return " ";
+    }
+    // check if in beta one or two mode. if invalid * marker then mark the barcode as invalid
+    if ((beta_2_mode && barcode[11] !== "2") || (!beta_2_mode && barcode[11] !== "1")) {
```

Comment:
`###` and `*` aren't actually what we use to denote these parts of the barcode, they're just arbitrary symbols I used to make the distinction when I was writing down the general format of the barcode. Hamed would know what `###` is actually called and we should use that name in the code. `*` probably doesn't have a name, you can just refer to it as the last digit of the barcode